### PR TITLE
Link pages via primary key

### PR DESCRIPTION
**Pull request by richtier**
[in relation to this issue](https://github.com/torchbox/wagtail-markdown/issues/40)

Allows users to create "vanilla" looking markdown hyperlinks to pages and documents with primary keys:

| markdown | result |
|---------------|-----------|
| `[document link](doc:1)` | `<a href="/documents/1/the-document.pdf">document link</a>`
| `[page link](page:1)` | `<a href="/path/to/page-1/">page link</a>` |
| `image` | `image` |

An advantage of this is there is a smaller learning curve for users over use of custom `<:My page name|link title>`, and the UI knows how to code highlight the new links:

image

Notice also that we use primary key instead of title, as currently the links break if the title changes, or if `wagtail-modeltranslations` has activated a languae other than the default.

A few other changes:

- Removed table extension as it's provided already by `python-markdown`
- Replaced the custom page and document linking extensions, and extended the existing `python-markdown` `markdown.inlinepatterns.ImagePattern` and `markdown.inlinepatterns.LinkPattern`.
- Placed the database lookup logic in it's own class that can be configured, to allow the user to do things like this:

```
# lookup the page by slug instead of pk using `[page](slug:page-slug)` as in my use-case slug is preferable over primary key
class SlugObjectLookupNegotiator(ObjectLookupNegotiator):
    PAGE_LINK_PREFIX = 'slug:'

    @staticmethod
    def retrieve_page(lookup_field_value):
        return Page.objects.get(slug=lookup_field_value)


def render_markdown_sluggy(text):
    return render_markdown(
        text, object_lookup_negotiator=SlugObjectLookupNegotiator
    )
```
This approach should simplify mocking during testing too.

I will write tests for this before it gets merged. I am looking for the green light on this change before I write them though :) I have a validator that confirms the primary keys are corrrect. I will include those in this PR for completeness.

How does version numbering work?

### CODE REVIEW

File: wagtailmarkdown/mdx/tables/__init__.py [hunk 1]
```diff
@@ -1,102 +0,0 @@
-"""
```
Comment (tm-kn) [inline wagtailmarkdown/mdx/tables/__init__.py]: I feel like we can do the tables as a separate PR beforehand this one.

**Comment by tm-kn** [comment]
Having initial look I am not sure of retiring old ways... That will break people's ways. We could probably slowly retire them or add a switch to enable them. What do you think?

> How does version numbering work?
The goal is to have a 0.5 release at some point soon that would be "stable". I released alpha release for the sake of having it handy on PyPI, but I can't certify its stability.

**Comment by richtier** [comment]
> I am not sure of retiring old ways... That will break people's ways.

The current situation is if a page title changes the CMS user would have to go through every markdown field on every page to check if there is a link to the old page's title. This is a recipe for 404s, and a blocker for adopting this library.

I agree breaking changes are important considerations though and a way to reduce the stress of the breaking change could be a data migration or a management command that programmatically convert "title linking" to "pk linking".

As for user experience change: as this is not yet a stable release we can also justify this change in functionality.

Has there been a discussion on using semantic versioning to communicate breaking changes to the devs?

**Comment by richtier** [comment]
closing this due to inactivity

**Comment by jhoover4** [comment]
Did we decide not to incorporate this?

**Comment by pbspm01** [comment]
Hi,
+1 for this, any chance to have this?
